### PR TITLE
Unit tests for heretic knowledge

### PR DESCRIPTION
**Pull request by Djiq**
## About The Pull Request

Adds unit tests for heretic knowledge tree.

Only merge this after https://github.com/tgstation/tgstation/pull/55020 is merged

## Changelog
:cl:
code: No more shittery with inaccessible nodes getting merged. 
/:cl:

### CODE REVIEW

**Comment by Jared-Fogle** [comment]
run the tests on ur own machine first

just uncomment `#define UNIT_TESTS` and run in dream daemon

**Comment by optimumtact** [comment]
You should collect the reachable ones in a list that you can intersect with
`subtypesof(base paths) - (unreachables)`

so you can show what wasn't hit

**Comment by EdgeLordExe** [comment]
> You should collect the reachable ones in a list that you can intersect with
> `subtypesof(base paths) - (unreachables)`
> 
> so you can show what wasn't hit

good idea

File: code/modules/antagonists/eldritch_cult/eldritch_antag.dm [hunk 1]
```diff
@@ -32,9 +32,8 @@
 	var/mob/living/current = owner.current
 	if(ishuman(current))
 		forge_primary_objectives()
-		gain_knowledge(/datum/eldritch_knowledge/spell/basic)
-		gain_knowledge(/datum/eldritch_knowledge/living_heart)
-		gain_knowledge(/datum/eldritch_knowledge/codex_cicatrix)
+		for(var/EK in GLOB.heretic_start_knowledge)
+			gain_knowledge(EK)
```
Comment (Jared-Fogle) [inline code/modules/antagonists/eldritch_cult/eldritch_antag.dm:36]:
```suggestion
		for(var/eldritch_knowledge in GLOB.heretic_start_knowledge)
			gain_knowledge(eldritch_knowledge)
```

File: code/modules/unit_tests/heretic_knowledge.dm [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+	var/list/all_possible_knowledge = subtypesof(/datum/eldritch_knowledge) - blacklist
+
+	var/list/list_to_check = GLOB.heretic_start_knowledge.Copy()
+	var/i = 0
+	while(i < length(list_to_check))
+		var/datum/eldritch_knowledge/EK = allocate(list_to_check[++i])
+		for(var/Y in EK.next_knowledge)
+			if(Y in list_to_check)
+				continue
+			list_to_check += Y
```
Comment (Jared-Fogle) [inline code/modules/unit_tests/heretic_knowledge.dm:13]:
```suggestion
		var/datum/eldritch_knowledge/eldritch_knowledge = allocate(list_to_check[++i])
		for(var/next_knowledge in eldritch_knowledge.next_knowledge)
			if(next_knowledge in list_to_check)
				continue
			list_to_check += next_knowledge
```

**Review by Mothblocks — CHANGES_REQUESTED** [review]
Should have a comment explaining what the test actually *does* as well, it's a bit confusing for me just from reading the code.

**Comment by Jared-Fogle** [comment]
Still requesting a comment on what the test actually does--it's not obvious to me just by reading code.

**Comment by EdgeLordExe** [comment]
> Still requesting a comment on what the test actually does--it's not obvious to me just by reading code.

it checks all heretic knowledge nodes - excluding the ones which are unreachable on purpose - if players can reach them in game, if it finds a single node that is unreachable it throws an error

**Comment by Jared-Fogle** [comment]
Cool, now put that in a comment 😛

File: code/modules/unit_tests/heretic_knowledge.dm [hunk 3]
```diff
@@ -1,3 +1,7 @@
+/*
+it checks all heretic knowledge nodes - excluding the ones which are unreachable on purpose - if players can reach them in game,
+if it finds a single node that is unreachable it throws an error
+*/
```
Comment (Jared-Fogle) [inline code/modules/unit_tests/heretic_knowledge.dm:4]:
```suggestion
/// This test checks all heretic knowledge nodes - excluding the ones which are unreachable on purpose - and ensures players can reach them in game.
/// If it finds a node that is unreachable, it throws an error.
```

Comment (Jared-Fogle) [reply code/modules/unit_tests/heretic_knowledge.dm:4]: Last nitpick, then it's fine.

**Pull request merged**